tree/ntuple/v7/src/RFieldMeta.cxx

```diff
@@ -186,39 +217,134 @@ std::size_t ROOT::Experimental::RClassField::AppendImpl(const void *from)
+                                              DescriptorId_t classFieldId)
 {
-   // This can happen for added base classes or non-simple members.
-   if (GetOnDiskId() == kInvalidDescriptorId) {
-      return;
+   auto idSourceMember = desc.FindFieldId(memberName, classFieldId);
+   if (idSourceMember != kInvalidDescriptorId)
+      return idSourceMember;
+
+   for (const auto &subFieldDesc : desc.GetFieldIterable(classFieldId)) {
+      const auto subFieldName = subFieldDesc.GetFieldName();
+      if (subFieldName.length() > 2 && subFieldName[0] == ':' && subFieldName[1] == '_') {
+         idSourceMember = LookupMember(desc, memberName, subFieldDesc.GetId());
+         if (idSourceMember != kInvalidDescriptorId)
+            return idSourceMember;
+      }
    }
-   // Gather all known sub fields in the descriptor.
+
+   return kInvalidDescriptorId;
+}
+
+void ROOT::Experimental::RClassField::PrepareStagingArea(const std::vector<const TSchemaRule *> &rules,
+                                                         const RNTupleDescriptor &desc,
+                                                         const RFieldDescriptor &classFieldDesc)
+{
+   auto clName = classFieldDesc.GetTypeName();
+   if (classFieldDesc.GetTypeVersion() != GetTypeVersion()) {
+      TClass::GetClass(classFieldDesc.GetTypeName().c_str())->GetStreamerInfo(classFieldDesc.GetTypeVersion());
+      clName += std::string("@@") + std::to_string(classFieldDesc.GetTypeVersion());
+   }
+   auto cl = TClass::GetClass(clName.c_str());
+   R__ASSERT(cl);
+
+   std::size_t stagingAreaSize = 0;
+   for (const auto rule : rules) {
+      for (auto source : TRangeDynCast<TSchemaRule::TSources>(rule->GetSource())) {
+         auto [itr, isNew] = fStagingItems.emplace(source->GetName(), RStagingItem());
+         if (!isNew) {
+            // This source member has already been processed by another rule (and we only support one type per member)
+            continue;
+         }
+         RStagingItem &stagingItem = itr->second;
+
+         const auto memberFieldId = LookupMember(desc, source->GetName(), classFieldDesc.GetId());
+         if (memberFieldId == kInvalidDescriptorId) {
+            throw RException(R__FAIL(std::string("cannot find on disk rule source member ") + GetTypeName() + "." +
+                                     source->GetName()));
+         }
+         const auto &memberFieldDesc = desc.GetFieldDescriptor(memberFieldId);
+
+         auto memberType = source->GetTypeForDeclaration() + source->GetDimensions();
+         stagingItem.fField = Create("" /* we don't need a field name */, std::string(memberType)).Unwrap();
+         stagingItem.fField->SetOnDiskId(memberFieldDesc.GetId());
+
+         stagingItem.fOffset = cl->GetDataMemberOffset(source->GetName());
+         stagingAreaSize = std::max(stagingAreaSize, stagingItem.fOffset + stagingItem.fField->GetValueSize());
+         R__ASSERT(stagingAreaSize > 0);
```

Comment:
The `stagingAreaSize` is already calculated and we should have at the end `stagingAreaSize == cl->Size();`

Reply:
We may skip rules, so we'll check for `stagingAreaSize <= fStagingClass->Size()`